### PR TITLE
fix: optimized protobuf access for performance

### LINKED ISSUE #145: Batches in >=2.0.0 are significantly slower
I believe there is a change of API or a performance issue when using batches with 2.1.0.

#### Environment details

  - OS type and version: Windows 10
  - Python version: 3.8.7
  - pip version: 20.3.3
  - `google-cloud-datastore` version: 2.1.0
  - `google-api-core` version: 1.24.1

#### Steps to reproduce

  1. Use DatastoreClient::put_multi with 500 entities inside a batch

#### Code example

```python
with client.batch():
    client.put_multi(chunk)
```

#### Performance
When using 1.15.0, my code can handle 9 batches of 500 entities in about 1-2 seconds while 2.1.0 takes literal minutes.
Disclaimer: I use the datastore emulator

**EDIT**: It can be reproduced in 2.0.0

### LINKED ISSUE #150: Too slow query fetch
Hi, team!

Imagine we have an entity Order with `customer_id` (str, uuid4), `created_at` (datetime) and other fields. I want to get all orders for specified customer and time range. 

The query returns about 1500 records.
The problem: it takes about 7-8 seconds!

#### Environment details
  - OS type and version: Docker image `python:3.8-slim`
  - Python version: Python 3.8.6
  - pip version: 20.2.2
  - `google-cloud-datastore` version: 2.0.1

#### Datastore indexes:
- kind: Order
  properties:
  - name: customer_id
  - name: created_at
    direction: asc

- kind: Order
  properties:
  - name: customer_id
  - name: created_at
    direction: desc

#### Code example

```python
query = db_client.query(
    kind='Order',
    filters=(
        ('customer_id', '=', customer_id),
        ('created_at', '>=', datetime_from),
        ('created_at', '<=', datetime_to),
    ),
)
result = list(query.fetch())
```

What I do wrong or can you suggest how to speed up this query?
Thanks!

### PR DESCRIPTION
**Pull request by craiglabenz**
More efficiently uses proto-plus wrappers, as well as inner protobuf attribute access, to greatly reduce the performance costs seen in version 2.0.0 (which stemmed from the introduction of proto-plus).

The size of the performance improvement scales with the number of attributes on each Entity, but in general, speeds once again closely approximate those from 1.15.

Fixes #145
Fixes #150

### CODE REVIEW

File: google/cloud/datastore/helpers.py [hunk 1]
```diff
@@ -119,11 +124,12 @@ def entity_from_protobuf(pb):
...
-    if not getattr(pb, "_pb", False):
-        # Coerce raw pb type into proto-plus pythonic type.
-        proto_pb = entity_pb2.Entity(pb)
-        pb = pb
+    if not isinstance(pb, entity_pb2.Entity):
+        # Coerce a dictionary into a raw protobuf
+        if not isinstance(pb, message.Message):
+            pb = entity_pb2.Entity.pb(**pb)
+        # Wrap the constructed protobuf with proto-plus dressing
+        proto_pb = entity_pb2.Entity.wrap(pb)
```
Comment (tseaver) [inline google/cloud/datastore/helpers.py:132]: We need to have test coverage for all three cases (dict, message, entity wrapper).

**Review by tseaver — CHANGES_REQUESTED** [review]

File: google/cloud/datastore/helpers.py (hunk 1 above)
Comment (craiglabenz) [reply google/cloud/datastore/helpers.py:132]: I'm surprised that untested line didn't get flagged by CI. I'll get on this.

Comment (craiglabenz) [reply google/cloud/datastore/helpers.py:132]: Test added ✅

**Comment by craiglabenz** [comment]
There are a few other untested code blocks, but I can't see the CI results for past PRs to confirm whether they were pre-existing.

**Review by tseaver — APPROVED** [review]

**Pull request merged**